### PR TITLE
[WIP] Adding Batch support for TWS Operator

### PR DESCRIPTION
**Pull request by ericm-db**
### What changes were proposed in this pull request?

We are allowing batch queries to use and define the `TransformWithState` operator, which was initially introduced for streaming.

### Why are the changes needed?

This is needed to keep up the parity between streaming and batch APIs, since we want everything supported in streaming to be supported in batch, as well. 

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Added unit tests that use the TransformWithState operator with a batch query.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/RemoveRedundantProjects.scala [hunk 1]
```diff
@@ -38,6 +38,7 @@ import org.apache.spark.sql.internal.SQLConf
  */
 object RemoveRedundantProjects extends Rule[SparkPlan] {
   def apply(plan: SparkPlan): SparkPlan = {
+    // try calling SparkPartialRule here, add flag to enable/disable
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/RemoveRedundantProjects.scala]: nit: should we remove this ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 2]
```diff
@@ -742,6 +742,7 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
           batchTimestampMs = None,
           eventTimeWatermarkForLateEvents = None,
           eventTimeWatermarkForEviction = None,
+          isStreaming = true,
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala:745]: This won't overwrite the batch rewrite earlier correct ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StatefulProcessorHandleImpl.scala [hunk 3]
```diff
@@ -68,21 +68,25 @@ class QueryInfoImpl(
...
+class StatefulProcessorHandleImpl(store: StateStore, runId: UUID, isStreaming: Boolean)
   extends StatefulProcessorHandle with Logging {
   import StatefulProcessorHandleState._
 
   private def buildQueryInfo(): QueryInfo = {
     val taskCtxOpt = Option(TaskContext.get())
     // Task context is not available in tests, so we generate a random query id and batch id here
-    val queryId = if (taskCtxOpt.isDefined) {
+    val queryId = if (!isStreaming) {
+      UUID.randomUUID().toString
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StatefulProcessorHandleImpl.scala]: Maybe just create a static UUID here. basically having a different one per task might be misleading ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala [hunk 4]
```diff
@@ -152,6 +154,8 @@ case class TransformWithStateExec(
   override protected def doExecute(): RDD[InternalRow] = {
     metrics // force lazy init at driver
 
+    // populate stateInfo if this is a streaming query
+    val stateInfo = getStateInfo
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala]: Where is this used ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala [hunk 5]
```diff
@@ -171,3 +176,45 @@ case class TransformWithStateExec(
...
+    val shufflePartitions = child.session.sessionState.conf.numShufflePartitions
+    val statefulOperatorStateInfo = StatefulOperatorStateInfo(
+      Utils.createTempDir().getAbsolutePath,
+      queryRunId = UUID.randomUUID(),
+      operatorId = 0,
+      storeVersion = 0,
+      numPartitions = shufflePartitions
+    )
+
+    // Rewrite physical operator to TransformWithStateForBatchExec
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala]: Should we remove this line ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StatefulProcessorHandleImpl.scala [hunk 6]
```diff
@@ -68,21 +68,25 @@ class QueryInfoImpl(
  * track of valid transitions as various functions are invoked to track object lifecycle.
  * @param store - instance of state store
  */
-class StatefulProcessorHandleImpl(store: StateStore, runId: UUID)
+class StatefulProcessorHandleImpl(store: StateStore, runId: UUID, isStreaming: Boolean)
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StatefulProcessorHandleImpl.scala]: We can have default value for `isStreaming` as true here ? That way the tests won't need to change ?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateSuite.scala [hunk 7]
```diff
@@ -131,25 +131,34 @@ class TransformWithStateSuite extends StateStoreMetricsTest
       )
     }
   }
+
```
Comment (anishshri-db) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateSuite.scala:134]: Nit: extra newline ?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateSuite.scala [hunk 8]
```diff
@@ -131,25 +131,34 @@ class TransformWithStateSuite extends StateStoreMetricsTest
...
-          TimeoutMode.NoTimeouts(),
-          OutputMode.Append())
-        .write
-        .format("noop")
-        .mode(SaveMode.Append)
-        .save()
-    }
-    assert(ex.isInstanceOf[AnalysisException])
-    assert(ex.getMessage.contains("not supported"))
+  test("transformWithState - batch should not fail") {
```
Comment (anishshri-db) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateSuite.scala]: We can just remove this ? Its effectively the same as test above right ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala (hunk 2 above)
Comment (ericm-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala:745]: Yeah they go through two different paths

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateSuite.scala [hunk 9]
```diff
@@ -19,7 +19,7 @@ package org.apache.spark.sql.streaming
 
 import org.apache.spark.SparkException
 import org.apache.spark.internal.Logging
-import org.apache.spark.sql.{AnalysisException, Encoders, SaveMode}
+import org.apache.spark.sql.SaveMode
```
Comment (anishshri-db) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateSuite.scala]: I guess `Encoders` we still need to keep ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StatefulProcessorHandleImpl.scala (hunk 3 above)
Comment (ericm-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StatefulProcessorHandleImpl.scala]: You mean set it to the same value each time?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateSuite.scala (hunk 9 above)
Comment (ericm-db) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateSuite.scala]: I don't think we do?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StatefulProcessorHandleImpl.scala [hunk 10]
```diff
@@ -68,21 +68,26 @@ class QueryInfoImpl(
...
   extends StatefulProcessorHandle with Logging {
   import StatefulProcessorHandleState._
 
   private def buildQueryInfo(): QueryInfo = {
     val taskCtxOpt = Option(TaskContext.get())
     // Task context is not available in tests, so we generate a random query id and batch id here
-    val queryId = if (taskCtxOpt.isDefined) {
+    // For batch queries, we populate the queryId manually
+    val queryId = if (!isStreaming) {
+      "00000000-0000-0000-0000-000000000000"
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StatefulProcessorHandleImpl.scala:80]: maybe store it as a val above ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala [hunk 11]
```diff
@@ -171,3 +175,44 @@ case class TransformWithStateExec(
...
+      outputMode,
+      outputObjAttr,
+      Some(statefulOperatorStateInfo),
+      Some(System.currentTimeMillis),
+      None,
+      None,
+      isStreaming = false,
+      child)
+  }
+
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala:217]: Nit: extra newline ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala [hunk 12]
```diff
@@ -57,6 +58,7 @@ case class TransformWithStateExec(
     batchTimestampMs: Option[Long],
     eventTimeWatermarkForLateEvents: Option[Long],
     eventTimeWatermarkForEviction: Option[Long],
+    isStreaming: Boolean = true,
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala:61]: Here I feel like we should require a value to be passed ? i.e. don't set a default value

**Pull request closed without merging**